### DISCUSSION TITLE
Suggestion for new feature  - Range Rule

### DISCUSSION THREAD

**Discussion by jetfreakk** [adobe/regola#10] [Ideas]
### Expected Behaviour

Range Validation: The RangeRule class will be specifically designed for validating whether a given fact falls within a specified numeric range with different inclusivity options. It's well-suited for scenarios where we have predefined expectations about the valid range of values for a particular fact.

### Actual Behaviour

Currently Range Rule is not supported.

**Comment by sic2** [comment]
@jetfreakk thanks for this suggestion. Could you expand on this concept? 
It would be good to provide some examples (as shown in the README) and list also the operations that would be supported (out of the existing ones and/or new ones).

Thanks

**Comment by jetfreakk** [comment]
Range Rule
The "Range Rule" will be used to evaluate whether a fact within a specified numeric range meets certain inclusivity criteria. 
It checks if a given numeric fact falls within the defined range, considering whether the minimum and maximum values are included or excluded from the range. It will extend KeyBasedRule

Example:

```
{
  "type": "RANGE",
  "key": "age",
  "min": 18,
  "max": 30,
  "inclusiveMin": true,
  "inclusiveMax": false
}
```


Key | Fact | Min | Max | InclusiveMin | InclusiveMax | Result 
-- | -- | -- | -- | -- | -- | -- 
"age" | 25 | 18 | 30 | true | false | VALID 
"age" | 30 | 18 | 30 | true | false | INVALID 
"age" | 18 | 18 | 30 | true | false | VALID 
"age" | 17 | 18 | 30 | true | false | INVALID 
"age" | 31 | 18 | 30 | true | false | INVALID 


I have also cloned regola and implemented a first draft of the same if that helps.

**Comment by sic2** [comment]
Some ideas...

Rather than have inclusideMin/inclusiveMax, might be good to define a set of operations to be consistent with the other rules.

In the future, we could potentially consider having other operations too. For example, one that compares ranges and determines if a range included, overlaps, etc...

A limitation of this proposal is that it works for integers only. Should it work for alphanumerical values too? What about doubles?

Feel free to share a fork of your implementation. I will be happy to have a quick look ahead of any PR.

**Comment by jetfreakk** [comment]
That sounds like a good idea too. 

I am assuming by some mentioned operations you mean the following (Tried drafting some examples of the operations)

| Key   |   Fact   |   Min Range   |   Max Range   |   Operation   |   Other Range Key   |   Result |
|------|--------|--------------|---------------|-------------|--------------------|------|
| range1 | {min: 5, max: 25} | 10 | 20 | INCLUDES | range2 | VALID |
| range1 | {min: 10, max: 20} | 5 | 25 | OVERLAPS | range2 | VALID |
| range1 | {min: 5, max: 10} | 15 | 20 | DOES_NOT_OVERLAP | range2 | VALID |


I have made use of "Numbers" so it should support Integers and Doubles. I don't see the use case of alphanumeric for this case. (Not sure how alphanumeric will come into picture here)

Sure,here is the fork. 

https://github.com/jetfreakk/regola/blob/main/src/main/java/com/adobe/abp/regola/rules/RangeRule.java

**Comment by jetfreakk** [comment]
Additional Note :- The Above example of Operations is Range to Range comparision (Overlapping, sub-range, etc) . The one I was thinking about is more on the lines of a fact lying inside a rule and the fact being a Number (Example, 5 lies in [5,10].

**Comment by sic2** [comment]
@jetfreakk can you open a PR from your fork so I can start reviewing your proposals? Thanks

**Comment by sic2** [comment]
Given our recent conversations offline, I have been giving some more thought about this.

As for the other rules, we want a similar syntax (see [here](https://github.com/adobe/regola?tab=readme-ov-file#value-based-rules)): `fact OPERATOR value/s`.

First of all, we need to define what a range is.
For some comparable type T:
- min, max: all values of type T from min to max inclusive
- min_exclusive, max_exclusive: all values of type T from min to max inclusive
- min, max, step: values from min to max at intervals defined by step


For a `RangeRule` this could be:
| Fact | Operator | Keys | Result |
| -- | -- | -- | -- |
| 10 | BETWEEN | min: 5, max: 20 | VALID |
| 10 | BETWEEN | min: 10, max: 20 | VALID | 
| 10 | BETWEEN | min_exclusive: 10, max: 20 | INVALID | 
| alpha | BETWEEN | min: a, max: beta | VALID |
| 10 | IN | min: 5, max: 20, step: 1 | VALID |
| 10 | IN | min: 5, max: 20, step: 2 | INVALID |
| 10 | IS_BEFORE | min: 15, max: 20 | VALID |
| 10 | IS_AFTER | min: 15, max: 20 | INVALID |

and so on.

The Range class from ApacheCommons is a good inspiration for this: https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/Range.html

The table above assumes that a fact is an individual value, but we could also have the fact being:
- a list of values 
- a range (then operators like OVERLAPS, IN will have different meanings)


I think as a first phase, we can start with a simpler version of the RangeRule and then build from that. 
We can mark this rule as `experimental` so that any clients using regola will know that this contract might be changing.

**Reply by jetfreakk** [reply]
Hi @sic2 , Great suggestions and thanks for helping out here.

I have created an initial draft PR for the same. Please do have a look and let me know your thoughts.  

PR : https://github.com/adobe/regola/pull/24

I have not added Test Cases or changed README for now. I thought it would be better once we agree upon the implementation

**Reply by sic2** [reply]
@jetfreakk linking back to your PR and my latest comment: https://github.com/adobe/regola/pull/24#issuecomment-2681809344, I do think we should have a basic README change and some test cases. 
We can be a bit more relaxed on what we release as this is experimental, but I believe there still needs to be some basic guardrails in place.

**Comment by sic2** [comment]
RangeRule released as part of regola v0.0.24: https://github.com/adobe/regola/releases/tag/regola-0.0.24
